### DISCUSSION TITLE
Issue with expected Error type mismatch

### DISCUSSION THREAD

**Discussion by DavidPSmavid** [hyperium/hyper#3674] [Q&A]
I have an issue where a function is returning a response with a `std::io::Error` when I would like for that error to be a `hyper::Error`. 

The idea is for the function to be able to return an HTML page on a GET request. Where I'm running into a problem is it looks like the `BoxBody` is trying to return `std::io::Error` in the Option by default.

```Rust
async fn serve_html(filename: &str) -> Result<Response<BoxBody<Bytes, hyper::Error>>, hyper::Error> {
    let file = File::open(filename).await;
    if file.is_err() {
        eprintln!("ERROR: Unable to open file");
        return Ok(not_found();
    }
    
    let file = file.unwrap();
    
    let reader_stream = ReaderStream::new(file);
    
    let stream_body = StreamBody::new(reader_stream.map_ok(Frame::data));
    let boxed_body = stream_body.boxed();
    
    let response = Response::builder()
        .status(StatusCode::OK)
        .body(boxed_body)
        .unwrap();
    
    Ok(response)
```

I'd like to do this in a way that returns a `hyper::Error` since the function it's feeding into is expecting it, so if my approach won't get me there I'd like to know. 

```Rust
pub async fn router (
    req: Request<hyper::body::Incoming>,
) -> Result<Response<BoxBody<Bytes, hyper::Error>>, hyper::Error> {
    match (req.method(), req.uri().path()) {
        (&Method::GET,  "/") => serve_html("src/static/upload.html").await,
        (&Method::POST, "/submit-upload") => upload::upload_file(req).await,
        _ => serve_html("src/static/404.html").await,
    }
}
```

Any suggestions would be much appreciated. Thanks in advance!

**Comment by seanmonstar** [comment]
You mean that that _body_ has an error type of `io::Error`. That makes sense, since reading from a `File` could encounter `io::Error`s. You can't convert that into a `hyper::Error`, since they are different.

It would help to learn a little about what the error in a body will do. If after returning some response headers, you start reading the file, and part way through it has an error, you'll want to signal that the response body cannot be completed. So you return an error. hyper will basically abort the response according to the way the HTTP protocol allows (depends on if its HTTP/1 or 2). But hyper doesn't really care what the exact body error is.

So, you could just make a `struct StreamingError;`, and then using something like `map_err(|err| { debug!("response body error: {}", err); StreamingError})`.

**Reply by DavidPSmavid** [reply]
Thanks for the reply, I'm new to Rust and am learning a lot. Can I take the error message from the `std::io::Error` and construct a new `hyper::Error` from it?

**Reply by seanmonstar** [reply]
No, `hyper::Error`s are purposefully only created inside the library. But, you shouldn't need to construct one. Anywhere you see something asking for a `hyper::Error`, it can actually be any error type. The error type you return from your service, and the error type in your body, they can both be any type you want (that implements `std::error::Error`).

**Reply by DavidPSmavid** [reply]
So if I do something like `let file = File::open(filename).await?;`, then I end up with the error
```Rust
`?` couldn't convert the error to `hyper::Error`
the question mark operation (`?`) implicitly performs a conversion on the error value using the `From` trait
the trait `From<hyper::error::Parse>` is implemented for `hyper::Error`
for that trait implementation, expected `hyper::error::Parse`, found `std::io::Error`
required for `Result<Response<BoxBody<hyper::body::Bytes, hyper::Error>>, hyper::Error>` to implement `FromResidual<Result<Infallible, std::io::Error>>`
```

Also, if I try expecting a `Box<dyn std::error::Error>` instead (like below), I get a different error
```Rust
async fn serve_html(filename: &str) -> Result<Response<BoxBody<Bytes, Box<dyn std::error::Error>>>, Box<dyn std::error::Error>> {
    let file = File::open(filename).await?;

    let reader_stream = ReaderStream::new(file);

    let stream_body = StreamBody::new(reader_stream.map_ok(Frame::data));
    let boxed_body = stream_body.boxed();

    let response = Response::builder()
        .status(StatusCode::OK)
        .body(boxed_body)
        .unwrap();

    Ok(response)
}
```

```Rust
expected struct `Response<BoxBody<_, Box<dyn std::error::Error + 'static'>>>`
found struct `Response<BoxBody<_, std::io::Error>>`
```

Does that mean the Errors are incompatible?

**Reply by beroal** [reply]
I believe you need to add `.map_err(Box:new)` or something like this before your `?`. This question is more about Rust than about Hyper.
